File: traccia/replay/renderer.py
"""
TRACCIA Replay API - Renderer

Timeline → 多格式输出
"""

import json
from .timeline import Timeline


class TimelineRenderer:
    """Timeline 多格式渲染器"""

    def render(self, timeline: Timeline, format: str = "text") -> str:
        if format == "json":
            return self._render_json(timeline)
        elif format == "md" or format == "markdown":
            return self._render_markdown(timeline)
        else:
            return self._render_text(timeline)

    def _render_text(self, timeline: Timeline) -> str:
        """纯文本渲染"""
        lines = []
        lines.append(f"Session: {timeline.session_objective}")
        lines.append("─" * 50)

        if not timeline.entries:
            lines.append("(无事件)")
            return "\n".join(lines)

        for entry in timeline.entries:
            # 提取时间部分 HH:MM:SS
            time_str = self._format_time(entry.timestamp)
            evidence_info = f"Evidence: {entry.evidence_count}" if entry.evidence_count > 0 else ""
            lines.append(f"\n[{time_str}] {entry.title}")
            if entry.summary:
                lines.append(f"           {entry.summary}")
            if evidence_info:
                lines.append(f"           {evidence_info}")

        return "\n".join(lines)

    def _render_markdown(self, timeline: Timeline) -> str:
        """Markdown 渲染"""
        lines = []
        lines.append(f"# Session: {timeline.session_objective}")
        lines.append("")
        lines.append("## Timeline")
        lines.append("")

        if not timeline.entries:
            lines.append("*(无事件)*")
            return "\n".join(lines)

        for entry in timeline.entries:
            time_str = self._format_time(entry.timestamp)
            lines.append(f"### {time_str} — {entry.title}")
            lines.append("")
            lines.append(f"- **Type**: `{entry.event_type}`")
            lines.append(f"- **Actor**: `{entry.actor_id}`")
            if entry.summary:
                lines.append(f"- **Summary**: {entry.summary}")
            if entry.evidence_count > 0:
                lines.append(f"- **Evidence**: {entry.evidence_count}")
            lines.append("")

        return "\n".join(lines)

    def _render_json(self, timeline: Timeline) -> str:
        """JSON 渲染"""
        return json.dumps(timeline.to_dict(), indent=2, ensure_ascii=False)

    def _format_time(self, timestamp: str) -> str:
        """从 ISO 8601 时间戳提取 HH:MM:SS"""
        try:
            # 格式: 2026-06-08T12:00:01Z 或 2026-06-08T12:00:01.123456Z
            if "T" in timestamp:
                time_part = timestamp.split("T")[1]
                # 去掉毫秒和时区
                time_part = time_part.split(".")[0].replace("Z", "")
                return time_part
        except Exception:
            pass
        return timestamp

File: traccia-rust/crates/traccia-replay/src/renderer.rs
use crate::timeline::Timeline;

pub struct TimelineRenderer;

impl TimelineRenderer {
    pub fn new() -> Self {
        Self
    }

    pub fn render(&self, timeline: &Timeline, format: &str) -> String {
        match format {
            "json" => self.render_json(timeline),
            "md" | "markdown" => self.render_markdown(timeline),
            _ => self.render_text(timeline),
        }
    }

    fn render_text(&self, timeline: &Timeline) -> String {
        let mut lines = Vec::new();
        lines.push(format!("Session: {}", timeline.session_objective));
        lines.push("─".repeat(50));

        if timeline.entries.is_empty() {
            lines.push("(无事件)".to_string());
            return lines.join("\n");
        }

        for entry in &timeline.entries {
            let time_str = extract_time(&entry.timestamp);
            lines.push(format!("\n[{}] {}", time_str, entry.title));
            if !entry.summary.is_empty() {
                lines.push(format!("           {}", entry.summary));
            }
            if entry.evidence_count > 0 {
                lines.push(format!("           Evidence: {}", entry.evidence_count));
            }
        }

        lines.join("\n")
    }

    fn render_markdown(&self, timeline: &Timeline) -> String {
        let mut lines = Vec::new();
        lines.push(format!("# Session: {}", timeline.session_objective));
        lines.push(String::new());
        lines.push("## Timeline".to_string());
        lines.push(String::new());

        if timeline.entries.is_empty() {
            lines.push("*(无事件)*".to_string());
            return lines.join("\n");
        }

        for entry in &timeline.entries {
            let time_str = extract_time(&entry.timestamp);
            lines.push(format!("### {} — {}", time_str, entry.title));
            lines.push(String::new());
            lines.push(format!("- **Type**: {}", entry.event_type));
            lines.push(format!("- **Actor**: {}", entry.actor_id));
            if !entry.summary.is_empty() {
                lines.push(format!("- **Summary**: {}", entry.summary));
            }
            if entry.evidence_count > 0 {
                lines.push(format!("- **Evidence**: {}", entry.evidence_count));
            }
            lines.push(String::new());
        }

        lines.join("\n")
    }

    fn render_json(&self, timeline: &Timeline) -> String {
        serde_json::to_string_pretty(timeline).unwrap_or_default()
    }
}

impl Default for TimelineRenderer {
    fn default() -> Self {
        Self::new()
    }
}

fn extract_time(timestamp: &str) -> String {
    // 从 ISO 8601 提取 HH:MM:SS
    if let Some(t_pos) = timestamp.find('T') {
        let time_part = &timestamp[t_pos + 1..];
        let time_part = time_part.split('.').next().unwrap_or(time_part);
        let time_part = time_part.trim_end_matches('Z').trim_end_matches('+');
        time_part.to_string()
    } else {
        timestamp.to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::timeline::{Timeline, TimelineEntry};

    #[test]
    fn test_render_text() {
        let mut timeline = Timeline::new("sess-1", "测试会话");
        timeline.add_entry(TimelineEntry {
            timestamp: "2026-06-08T10:01:12Z".to_string(),
            event_id: "evt-1".to_string(),
            event_type: "read_file".to_string(),
            title: "读取文件".to_string(),
            actor_id: "agent_001".to_string(),
            evidence_count: 2,
            summary: "读取了 /data/test.csv".to_string(),
        });

        let renderer = TimelineRenderer::new();
        let output = renderer.render(&timeline, "text");

        assert!(output.contains("测试会话"));
        assert!(output.contains("读取文件"));
        assert!(output.contains("Evidence: 2"));
    }

    #[test]
    fn test_render_json() {
        let timeline = Timeline::new("sess-1", "测试");
        let renderer = TimelineRenderer::new();
        let output = renderer.render(&timeline, "json");

        assert!(output.contains("timeline_id"));
        assert!(output.contains("sess-1"));
    }
}
